src/stream/src/executor/temporal_filter.rs

```diff
@@ -0,0 +1,466 @@
+                // Remove the record from buffer.
+                let row = entry.remove();
+
+                // If there is higher condition, we need to check existence before yield.
+                if let Some(data_chunk) =
+                    data_chunk_builder.append_one_row_from_datums(row.values())
+                {
+                    // When the chunk size reaches its maximum, we construct a
+                    // stream chunk
+                    let ops = vec![Op::Delete; data_chunk.capacity()];
+                    ret.push(StreamChunk::from_parts(ops, data_chunk));
+                }
+
+                self.state_table.delete(row);
+            }
+        }
+        if let Some(data_chunk) = data_chunk_builder.consume_all() {
+            let ops = vec![Op::Delete; data_chunk.capacity()];
+            ret.push(StreamChunk::from_parts(ops, data_chunk));
+        }
+        Ok(ret)
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn into_stream(mut self) {
+        let time_col_data_type = self.schema().fields()[self.time_col_idx].data_type();
+        let window_size = *self.window_size.as_interval();
+        let dynamic_cond = move |time_millis: u64| -> Result<BoxedExpression, ExprError> {
+            new_binary_expr(
+                Type::GreaterThanOrEqual,
+                DataType::Boolean,
+                Box::new(InputRefExpression::new(
+                    time_col_data_type.clone(),
+                    self.time_col_idx,
+                )),
+                Box::new(LiteralExpression::new(
+                    DataType::Timestamp,
+                    Some(ScalarImpl::NaiveDateTime(timestamp_interval_sub::<
+                        NaiveDateTimeWrapper,
+                        IntervalUnit,
+                        NaiveDateTimeWrapper,
+                    >(
+                        NaiveDateTimeWrapper::new(NaiveDateTime::from_timestamp(
+                            (time_millis / 1000) as i64,
+                            (time_millis % 1000 * 1_000_000) as u32,
+                        )),
+                        window_size,
+                    )?)),
+                )),
+            )
+        };
+
+        let mut input = self.input.take().unwrap().execute();
+
+        // Consume the first barrier message and initialize state table.
+        let barrier = expect_first_barrier(&mut input).await?;
+        self.state_table.init_epoch(barrier.epoch);
+        let mut time_millis = Epoch::from(barrier.epoch.curr).as_unix_millis();
+        let mut sync_expr = None;
+
+        // The first barrier message should be propagated.
+        yield Message::Barrier(barrier);
+
+        // Fill the buffer on initialization. If the executor has recovered from a failed state, the
+        // buffer should be refilled by its previous data.
+        let vnode_bitmap = self.state_table.vnode_bitmap().to_owned();
+        self.fill_buffer(None, &vnode_bitmap).await?;
```

Comment:
Basically, I think the initial sync should be performed here. Not when you receive the first chunk.

Reply:
> Basically, I think the initial sync should be performed here. Not when you receive the first chunk.

This is deliberate because the second barrier might be an update barrier, which means that we are not allowed emit any chunks.